packages/core/src/features/st-scope.ts

```diff
@@ -0,0 +1,59 @@
+import { createFeature } from './feature';
+import { parseSelectorWithCache, scopeNestedSelector } from '../helpers/selector';
+import type { ImmutablePseudoClass } from '@tokey/css-selector-parser';
+import * as postcss from 'postcss';
+import type { SRule } from '../deprecated/postcss-ast-extension';
+
+export const diagnostics = {
+    MISSING_SCOPING_PARAM() {
+        return '"@st-scope" missing scoping selector parameter';
+    },
+};
+
+// HOOKS
+
+export const hooks = createFeature<{ IMMUTABLE_SELECTOR: ImmutablePseudoClass }>({
+    analyzeAtRule({ context, atRule, analyzeRule }) {
+        if (!isStScopeStatement(atRule)) {
+            return;
+        }
+        if (!atRule.params) {
+            context.diagnostics.warn(atRule, diagnostics.MISSING_SCOPING_PARAM());
+        }
+        analyzeRule(
+            postcss.rule({
+                selector: atRule.params,
+                source: atRule.source,
+            }),
+            {
+                isScoped: true,
+            }
+        );
+        context.meta.scopes.push(atRule);
+    },
+    prepareAST({ node, toRemove }) {
+        if (isStScopeStatement(node)) {
+            flattenScope(node);
+            toRemove.push(() => node.replaceWith(node.nodes || []));
```

Comment:
maybe we want to change the name toRemove this is doing the replacement

Reply:
not sure of what's a better name for this, anyhow it should be dismantled into the transformer flow in the future, so the API should be removed eventually